src/Lang/Surface.ml

```diff
@@ -247,15 +282,24 @@ and expr_data =
       other have no effect. *)
 
 (** Explicit instantiation of named parameters in polymorphic expression *)
-and inst = (name * expr) node
+and inst = inst_data node
+and inst_data =
+  | IType   of tvar * type_expr
+    (** Explicit instantiation of a type variable *)
+  | IVal    of name * expr
+    (** Explicit instantiation of a value-level name *)
+  | IModule of module_name path
+    (** Explicit instantiation, that takes values from given module *)
+  | IOpen
+    (** Explicit instantiation, that takes values from the environment *)
 
 (** Local definitions *)
 and def = def_data node
 and def_data =
-  | DLetId of ident * expr
+  | DLetId of is_public * ident * expr
     (** Let definition: monomorphic or polymorphic, depending on effect *)
 
-  | DLetFun of ident * named_type_arg list * named_arg list * expr
```

Comment:
Polymorphic lambdas in argument position allow any order of type/named value parameters, so I think so should let-definitions for consistency.

Reply:
Polymorphic lambdas are written in a context, where the scheme is known. Allowing any `named_pattern` here, leads to some problematic cases, like:
```
let foo {module M} = M.bar M.baz
```
But maybe such programs can be rejected by the type-checker.